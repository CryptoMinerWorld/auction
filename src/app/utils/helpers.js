export const OxToLowerCase = (id) => 
      id.split('')
      .map(item => (typeof item === 'string' ? item.toLowerCase() : item))
      .join('');