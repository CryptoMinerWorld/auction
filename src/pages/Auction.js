import React, { Component } from 'react';
import MobileHeader from '../components/MobileHeader';
import AuctionImage from '../components/AuctionImage';
import styled from 'styled-components';
import AuctionBox from '../components/AuctionBox';
import DescriptionBox from '../components/DescriptionBox';
import ProgressMeter from '../components/ProgressMeter';
import FAQ from '../components/FAQ';
import MailingList from '../components/MailingList';

const StickyHeader = styled.div`
  position: -webkit-sticky; /* Safari */
  position: sticky;
  top: 0;
  z-index: 3;
`;

class Auction extends Component {
  render() {
    let {
      currentPrice,
      minPrice,
      maxPrice,
      level,
      grade,
      rate,
      buyNow
    } = this.props;
    return (
      <div>
        <StickyHeader>
          <MobileHeader
            currentPrice={currentPrice}
            level={level}
            grade={grade}
            rate={rate}
          />
        </StickyHeader>
        <AuctionImage />
        <AuctionBox
          currentPrice={currentPrice}
          handleBuyNow={buyNow}
          level={level}
          grade={grade}
          rate={rate}
        />
        <DescriptionBox level={level} grade={grade} rate={rate} />
        <ProgressMeter
          currentPrice={currentPrice}
          minPrice={minPrice}
          maxPrice={maxPrice}
        />
        <FAQ />
        <MailingList />
      </div>
    );
  }
}

export default Auction;
