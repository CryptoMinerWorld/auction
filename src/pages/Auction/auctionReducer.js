import { AUCTION_DETAILS_RECEIVED} from "./auctionConstants"

export default function auctionReducer (state = {} , action) {
    
    if (action.type === AUCTION_DETAILS_RECEIVED){
        return action.payload
    }
    
    return state
    
}