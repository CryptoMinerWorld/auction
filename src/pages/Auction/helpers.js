export const isTokenForSale = (_contract, _tokenId) =>
  new Promise((resolve, reject) => {
    _contract.isTokenOnSale(_tokenId, (error, result) => {
      if (!error) resolve(result);
      else reject(error);
    });
  });

export const getAuctionDetails = (_contract, _tokenId) =>
  new Promise((resolve, reject) => {
    _contract.items(_tokenId, (error, result) => {
      if (!error) {
        resolve(result);
      } else reject(error);
    });
  });

// export const calcMiningRate = (gradeType, gradeValue) => {
//   switch (gradeType) {
//     case 1:
//       return gradeValue / 200000;
//     case 2:
//       return 10 + gradeValue / 200000;
//     case 3:
//       return 20 + gradeValue / 200000;
//     case 4:
//       return 40 + (3 * gradeValue) / 200000;
//     case 5:
//       return 100 + gradeValue / 40000;
//     case 6:
//       return 300 + gradeValue / 10000;
//     default:
//       return 300 + gradeValue / 10000;
//   }
// };

export const calcMiningRate = (gradeType, gradeValue) => ({
  1: gradeValue / 200000,
  2: 10 + gradeValue / 200000,
  3: 20 + gradeValue / 200000,
  4: 40 + (3 * gradeValue) / 200000,
  5: 100 + gradeValue / 40000,
  6: 300 + gradeValue / 10000
}[gradeType]);

export const getGemQualities = (_contract, _tokenId) =>
  new Promise((resolve, reject) => {
    _contract.getProperties(_tokenId, (error, properties) => {
      if (!error) {
        const color = properties.dividedToIntegerBy(0x10000000000).toNumber();
        const level = properties
          .dividedToIntegerBy(0x100000000)
          .modulo(0x100)
          .toNumber();
        const gradeType = properties
          .dividedToIntegerBy(0x1000000)
          .modulo(0x100)
          .toNumber();
        const gradeValue = properties.modulo(0x1000000).toNumber();
        resolve([color, level, gradeType, gradeValue]);
      } else reject(error);
    });
  });
