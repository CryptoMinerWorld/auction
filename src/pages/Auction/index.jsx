import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Confetti from 'react-confetti'
import sizeMe from 'react-sizeme'
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import AuctionImage from '../../components/AuctionImage';
import AuctionBox from '../../components/AuctionBox';
import DescriptionBox from '../../components/DescriptionBox/index';
import FAQ from '../../components/FAQ';
import MailingList from '../../components/MailingList';
import './animations.css';
import rockBackground from '../../images/rockBackground.png';

const OverlapOnDesktopView = styled.div`
  @media (min-width: 64em) {
    position: absolute;
    bottom: 1em;
    left: 5em;
  }
`;

const RockOverlay = styled.div`
  background-image: url(${rockBackground});
  background-repeat: repeat;
  background-size: contain;
`;

const TopHighlight = styled.div`
  background: linear-gradient(to right, #e36d2d, #b91a78);
  border-radius: 4px 4px 0px 0px;
  height: 3px;
`;

const Auction = ({
  currentPrice,
  minPrice,
  maxPrice,
  level,
  grade,
  rate,
  buyNow,
  deadline,
  name,
  tokenId,
  redirectTo,
  showConfirm,
  color,
  sourceImage,
  story,
  size,
  releaseConfetti
}) => (
    <div>
      {releaseConfetti && <div style={{ position: 'absolute', top: 0, left: 0, width: '100%', height: '100%', zIndex: '999' }}>
        <Confetti  {...size} />
      </div>}
      <div className="bg-off-black ">

        <RockOverlay>
          <div className="relative mw9 center">
            <AuctionImage sourceImage={sourceImage} />
            <ReactCSSTransitionGroup
              transitionName="example"
              transitionAppear
              transitionAppearTimeout={5000}
              transitionEnterTimeout={5000}
              transitionLeaveTimeout={5000}
            >
              <AuctionBox
                currentPrice={currentPrice}
                minPrice={minPrice}
                maxPrice={maxPrice}
                deadline={deadline}
                handleBuyNow={buyNow}
                level={level}
                grade={grade}
                rate={rate}
                name={name}
                tokenId={tokenId}
                redirectTo={redirectTo}
                showConfirm={showConfirm}

              />
            </ReactCSSTransitionGroup>
          </div>
        </RockOverlay>
        <div className="bg-off-black">
          <TopHighlight />
          <div className="mw9 center relative-l">
            <DescriptionBox level={level} grade={grade} rate={rate} color={color} story={story} name={name} />
            <div className="w-50-l measure-wide-l">
              <OverlapOnDesktopView>
                <FAQ />
              </OverlapOnDesktopView>
            </div>
          </div>
        </div>
        <MailingList />
      </div>
    </div>
  );


export default sizeMe({
  monitorHeight: true,
  monitorWidth: true,
})(Auction);

Auction.propTypes = {
  size: PropTypes.shape({
    monitorHeight: PropTypes.bool,
    monitorWidth: PropTypes.bool
  }).isRequired,
  deadline: PropTypes.number.isRequired,
  currentPrice: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  minPrice: PropTypes.number.isRequired,
  maxPrice: PropTypes.number.isRequired,
  level: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  grade: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  rate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  buyNow: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  showConfirm: PropTypes.func.isRequired,
  tokenId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  redirectTo: PropTypes.string,
  color: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  sourceImage: PropTypes.string.isRequired,
  story: PropTypes.string.isRequired,
};

Auction.defaultProps = {
  redirectTo: '',

};