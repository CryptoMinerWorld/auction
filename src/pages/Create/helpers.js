// export const handleApproveGemTransfer = (
//   _contract,
//   _approvedContract,
//   _tokenId
// ) =>
//   new Promise((resolve, reject) => {
//     _contract.approve(_approvedContract, _tokenId, (error, result) => {
//       if (!error) resolve(_tokenId, result);
//       else reject(error);
//     });
//   });

export const ethToWei = eth => Number((eth * 1000000000000000000).toFixed(20));

export const daysToMilliseconds = days => Number((days * 86400000).toFixed(20));

export const daysToSeconds = days => Number(days * 86400);

