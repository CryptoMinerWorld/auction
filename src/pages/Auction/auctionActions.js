import {  AUCTION_DETAILS_RECEIVED} from './auctionConstants'
import {db} from '../../utils/firebase'
import store from '../../store'


export const getAuctionDetails = tokenId => db
 .doc(`auctions/${tokenId}`)
 .get()
 .then(doc => {
    console.log("poo", doc.data(), typeof tokenId);
   
    return store.dispatch({
     type:AUCTION_DETAILS_RECEIVED, 
     payload: doc.data()
    })

 })
 .catch(error => console.error('error', error))
    
 
 export const temp = () => console.log('temp')
 
 
 