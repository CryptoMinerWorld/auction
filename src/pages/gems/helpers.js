import fromExponential from 'from-exponential';
import {BigNumber} from 'bignumber.js';
import { db, storage } from '../../utils/firebase'


export const isTokenForSale = (_contract, _tokenId) =>
 
    _contract.methods.isTokenOnSale(_tokenId).call();


export const getAuctionDetails = (_contract, _tokenId) =>
    _contract.methods.items(_tokenId).call().then( result => {
      const {t0, t1, p0, p1} = result;
      return([t0, t1, p0, p1])
    })

export const getGemStory = (color, level) => {
      const type = {
          9: 'sapphire',
          10: 'opal',
          1: 'garnet',
          2: 'amethyst',
      }[color]
      const lvl = `lvl${level}`
      return db.doc(`gems/${type}`).get().then(doc => doc.data()[lvl])
  }


export const  getGemImage = (color, grade, level) => {

    const type = {
        9: 'Sap',
        10: 'Opa',
        1: 'Gar',
        2: 'Ame',
    }[color]

    const gradeType = {
        1: 'D',
        2: 'C',
        3: 'B',
        4: 'A',
        5: 'AA',
        6: 'AAA',
    }[grade]

    const sourceImage = `${type}-${level}-${gradeType}-4500.png`;

    return storage
        .ref(`gems512/${sourceImage}`)
        .getDownloadURL()
}


export const calcMiningRate = (gradeType, gradeValue) => ({
    1: gradeValue / 200000,
    2: 10 + gradeValue / 200000,
    3: 20 + gradeValue / 200000,
    4: 40 + (3 * gradeValue) / 200000,
    5: 100 + gradeValue / 40000,
    6: 300 + gradeValue / 10000
  }[gradeType]);


export const getGemQualities = (_contract, _tokenId) => 
_contract.methods
.getProperties(_tokenId)
.call()
.then( _properties => {

    // web3.utils.toBN(number)
const properties = new BigNumber(_properties)
        const color = properties.dividedToIntegerBy(0x10000000000).toNumber();
        const level = properties
          .dividedToIntegerBy(0x100000000)
          .modulo(0x100)
          .toNumber();
        const gradeType = properties
          .dividedToIntegerBy(0x1000000)
          .modulo(0x100)
          .toNumber();
        const gradeValue = properties.modulo(0x1000000).toNumber();
        return([color, level, gradeType, gradeValue]);
      })
  
      
 export const getPrice =  (_tokenId, _contract) =>  _contract.methods.getCurrentPrice(
        _tokenId).call()
        
export const nonExponential = (count) =>     fromExponential(Number(count) / 1000000000000000000)

export const calculateGemName = (providedGrade, providedTokenId) => {
    const gemType = {
      9: "Sapphire",
      10: "Opal",
      1: "Garnet",
      2: "Amethyst"
    }[providedGrade];
    return `${gemType} #${providedTokenId}`;
  };








// converts BigNumber representing Solidity uint256 into String representing Solidity bytes
const toBytes = uint256 => {
  let s = uint256.toString(16);
  const len = s.length;
  // 256 bits must occupy exactly 64 hex digits
  if (len > 64) {
    s = s.substr(0, 64);
  }
  for (let i = 0; i < 64 - len; i += 1) {
    s = `0${s}`;
  }
  return `0x${s}`;
};




  // @notice creates an auction
export const createAuctionHelper = async (
    _tokenId,
    _duration,
    _startPriceInWei,
    _endPriceInWei,
    _contract,
    _currentAccount
  ) => {
    // construct auction parameters
    const token = Number(_tokenId);
    const tokenId = new BigNumber(token);
    const t0 = Math.round(new Date().getTime() / 1000) || 0;
    const t1 = t0 + _duration;
    const p0 = _startPriceInWei;
    const p1 = _endPriceInWei;
    const two = new BigNumber(2);
  
    // convert auction parameters to bytecode for smart contract
    const data = toBytes(
      two
        .pow(224)
        .times(tokenId)
        .plus(two.pow(192).times(t0))
        .plus(two.pow(160).times(t1))
        .plus(two.pow(80).times(p0))
        .plus(p1)
    );
  
    // submit the auction
    const auctionSuccessfullyCreated = _contract.methods
      .safeTransferFrom(
        _currentAccount,
        process.env.REACT_APP_DUTCH_AUCTION,
        token,
        data
      )
      .send()
      .on("receipt", () => true)
      .catch(() => false);
  
    // get gem details
    const gemProperties = getGemQualities(_contract, _tokenId);
  
    // only proceed if auction was successfully created and you have gem details
    Promise.all([auctionSuccessfullyCreated, gemProperties])
      .then(result => {
        // destructure and format gem properties from result
        const [color, level, gradeType, gradeValue] = result[1];
        const grade = gradeType;
        const gemLevel = Number(level);
        const rate = Number(calcMiningRate(gradeType, gradeValue));
  
        // get the gem image and story since they are not stored on blockchain
        const image = getGemImage(color, gradeType, gemLevel);
        const story = getGemStory(color, gemLevel);
  
        // only proceed once you have gem image and story
        Promise.all([image, story])
          .then(([_image, _story]) => {
            const finalAuction = {
              id: _tokenId,
              minPrice: _startPriceInWei,
              maxPrice: _endPriceInWei,
              owner: _currentAccount,
              deadline: t1,
              gemImage: _image,
              story: _story,
              grade,
              gemLevel,
              color,
              rate
            };
  
            return finalAuction
          })
          .catch(err => console.error(err));
      })
      .catch(err => console.error(err));
  };
