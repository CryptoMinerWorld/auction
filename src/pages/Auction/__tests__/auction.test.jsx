import React from "react";
import {
  render,
  fireEvent,
  waitForElement,
  cleanup
} from "react-testing-library";
import "jest-dom/extend-expect";
import Auction from "..";
import { calculateGemName } from "../helpers";

import { calcMiningRate } from "../helpers";

const firebasemock = require("firebase-mock");

const mockfirestore = new firebasemock.MockFirestore();
const mockstorage = new firebasemock.MockStorage();

// @dev this automatically unmounts and cleanup DOM after the test is finished.
afterEach(cleanup);

// @dev this lets you set a deadline so that the time is easy to test
const someDate = new Date();
const numberOfDaysToAdd = 2;
someDate.setDate(someDate.getDate() + numberOfDaysToAdd);

describe.skip("Auction page tests", () => {
  // @dev this is all the test data, easy to configure in one place
  const props = {
    currentPrice: "1.323",
    minPrice: 0.8,
    maxPrice: 4.5,
    deadline: Math.round(someDate / 1000),
    level: 2,
    grade: 5,
    rate: 53,
    color: 10,
    name: "Amethyst #12345",
    buyNow: jest.fn(),
    showConfirm: jest.fn(),
    gemId: 12345,
    story: "hello",
    gemImage: "hello url",
    tokenId: 12345,
    redirectTo: "",
    auctionStartTime: 2000,
    auctionEndTime: 1000,
    sourceImage:
      "https://i.kym-cdn.com/photos/images/original/001/225/594/18a.gif",
    provider: true,
    currentTime: 1536908786405,
    currentAccount: "123"
  };

  test("Auction Page render correctly", async () => {
    const { container } = render(<Auction {...props} />);
    expect(container).toMatchSnapshot();
  });

  test("Buy now button triggers the modal with the correct gem Id and the buy Now function", async () => {
    // Arrange
    const { getByTestId } = render(<Auction {...props} />);

    const buyNowButton = await waitForElement(() =>
      getByTestId("buyNowButton")
    );
    // Act
    fireEvent.click(buyNowButton);

    // Assert
    expect(props.buyNow).toHaveBeenCalledTimes(1);
    expect(props.buyNow).toHaveBeenCalledWith(
      props.gemId,
      props.currentAccount
    );
  });

  test("If no metamask show modal, otherwise let people buy directly", async () => {
    const { getByTestId } = render(<Auction {...props} provider={false} />);

    const metaMask = props.web3;
    expect(metaMask).toBeFalsy();
    // Act
    const buyNowButton = await waitForElement(() =>
      getByTestId("buyNowButton")
    );
    fireEvent.click(buyNowButton);

    expect(props.showConfirm).toHaveBeenCalledTimes(1);
    expect(props.showConfirm).toHaveBeenCalledWith(
      props.buyNow,
      props.gemId,
      props.currentAccount
    );
  });

  test.skip("Countdown timer shows correct time", async () => {
    const { getByTestId } = render(<Auction {...props} />);

    const days = await waitForElement(() => getByTestId("daysLeft"));
    const hours = await waitForElement(() => getByTestId("hoursLeft"));
    const minutes = await waitForElement(() => getByTestId("minutesLeft"));
    const seconds = await waitForElement(() => getByTestId("secondsLeft"));

    expect(days).toHaveTextContent(1);
    expect(hours).toHaveTextContent(23);
    expect(minutes).toHaveTextContent(59);
    expect(seconds).not.toHaveTextContent(0);
  });

  test.skip("Countdown timer shows non-plural time descriptions (for example 1 hour vs 1 hours)", async () => {
    const { getByTestId } = render(<Auction {...props} />);

    const daysUnit = await waitForElement(() => getByTestId("daysUnit"));
    expect(daysUnit).not.toHaveTextContent("days");
    expect(daysUnit).toHaveTextContent("day");
  });

  test.skip("Progress bar shows correct start and end price", async () => {
    const { getByTestId } = render(<Auction {...props} />);

    const minPriceBit = await waitForElement(() => getByTestId("minPrice"));
    const maxPriceBit = await waitForElement(() => getByTestId("maxPrice"));
    const currentPriceBit = await waitForElement(() =>
      getByTestId("currentPrice")
    );

    expect(minPriceBit).toHaveTextContent(props.minPrice);
    expect(maxPriceBit).toHaveTextContent(props.maxPrice);
    expect(currentPriceBit).toHaveTextContent(props.currentPrice);
  });

  test.skip("Current price shows the correct price", async () => {
    const { getByTestId } = render(<Auction {...props} />);

    const currentPriceBit = await waitForElement(() =>
      getByTestId("currentPrice")
    );

    expect(currentPriceBit).toHaveTextContent(props.currentPrice);
  });

  test.skip("Check if the auction is still active, show bought or over", async () => {
    const { queryByText } = render(
      <Auction {...props} isTokenOnSale={false} />
    );
    const Button = queryByText("submit");
    expect(Button).toBeInTheDocument();
  });

  test.skip("auction tell you an auction is over of the id does not exist", () => {
    expect(true).toBeFalsy();
  });

  test.skip("Current price depreciates over time", async () => {
    // I don't knwo how to test this since the time deprecation is happening on the contract
    // jest.advanceTimersByTime(1000); might be useful somewhere
    // https://jestjs.io/docs/en/timer-mocks.html#advance-timers-by-time
  });
  test.skip("Current price does not continue to depreciate after the deadline", async () => {
    // Again, I don't knwo how to test this since the time deprecation is happening on the contract
    // jest.advanceTimersByTime(1000); might be useful somewhere
    // https://jestjs.io/docs/en/timer-mocks.html#advance-timers-by-time
  });

  test("calcMiningRate accurately calculates the mining rate", () => {
    expect(calcMiningRate(5, 700203)).toEqual(117.505075);
    expect(calcMiningRate(1, 659700)).toEqual(3.2985);
  });

  test.skip("auction lets people buy after a deadline is passed", () => {
    // Again, I don't knwo how to test this since the time deprecation is happening on the contract
    // jest.advanceTimersByTime(1000); might be useful somewhere
    // https://jestjs.io/docs/en/timer-mocks.html#advance-timers-by-time
  });

  test.skip("The gem name renders correctly", async () => {
    const { getByTestId } = render(<Auction {...props} />);
    const name = await waitForElement(() => getByTestId("gemName"));
    expect(name).toHaveTextContent(props.name);

    const testname = calculateGemName(2, 12355);
    expect(testname).toEqual("Amethyst #12355");
  });

  test.skip("Auction form can be submitted with 0 as the final price", async () => {});
  test.skip("When you pourchase a gem the auction is no longer accessible without a refresh", async () => {});
  test.skip("When you purchase a gem the redirect waits for transfer event", async () => {});
});
