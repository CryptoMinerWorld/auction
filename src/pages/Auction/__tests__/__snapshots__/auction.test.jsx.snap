// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auction page tests Auction Page render correctly 1`] = `
<div>
  <div
    style="width: 100%; height: 100%;"
  />
</div>
`;
